go-controller/pkg/libovsdbops/db_index.go

```diff
@@ -0,0 +1,360 @@
+	// May be used by controllers to store e.g. namespace+name of the object.
+	SecondaryIndexKey = types.OvnK8sPrefix + "/name"
+	// PrimaryIDKey will be used as a primary index, that is unique for every db object,
+	// and can be built based on the combination of all the other ids.
+	PrimaryIDKey = types.OvnK8sPrefix + "/id"
+)
+
+type dbObjType int
+
+const (
+	addressSet dbObjType = iota
+	// ACL here
+)
+
+type controllerType int
+
+const (
+	defaultController controllerType = iota
+	// secondaryLayer3Controller here
+)
+
+type ownerType string
+
+const (
+	// owner types
+	EgressFirewallDNSOwnerType ownerType = "EgressFirewallDNS"
+	EgressQoSOwnerType         ownerType = "EgressQoS"
+	NetworkPolicyOwnerType     ownerType = "NetworkPolicy"
+	NamespaceOwnerType         ownerType = "Namespace"
+	HybridNodeRouteOwnerType   ownerType = "HybridNodeRoute"
+	EgressIPOwnerType          ownerType = "EgressIP"
+	EgressServiceOwnerType     ownerType = "EgressService"
+
+	// owner extra IDs
+	EgressQosAddrSetPriorityKey = "priority"
+	PolicyDirectionID           = "direction"
+	GressIdxID                  = "gress-index"
+	AddressSetIPFamilyID        = "ip-family"
+)
+
+// dbIndexes is used to track indexes used by different controllers and objects.
+// ownerType is used directly in the ExternalIDs with OwnerObjectType key.
+// The same ownerType may be used by different (dbObjType, controllerType) combinations,
+// therefore this map will make sure the same key is not used twice by any (dbObjType, controllerType).
+// To create a unique id, values for listed keys will be joined with colon in the order they are listed here,
+// to make unique ID easier to read, sort ids in the narrowing-scope order.
+var dbIndexes = map[dbObjType]map[controllerType]map[ownerType][]string{
+	addressSet: {
+		defaultController: {
+			EgressFirewallDNSOwnerType: {
+				// dnsName
+				SecondaryIndexKey,
+				AddressSetIPFamilyID,
+			},
+			HybridNodeRouteOwnerType: {
+				// nodeName
+				SecondaryIndexKey,
+				AddressSetIPFamilyID,
+			},
+			EgressQoSOwnerType: {
+				// namespace
+				SecondaryIndexKey,
+				// priority
+				EgressQosAddrSetPriorityKey,
+				AddressSetIPFamilyID,
+			},
+			NetworkPolicyOwnerType: {
+				// namespace_name
+				SecondaryIndexKey,
+				// egress or ingress
+				PolicyDirectionID,
+				// gress rule index
+				GressIdxID,
+				AddressSetIPFamilyID,
+			},
+			NamespaceOwnerType: {
+				// namespace
+				SecondaryIndexKey,
+				AddressSetIPFamilyID,
+			},
+			EgressIPOwnerType: {
+				// cluster-wide address set name
+				SecondaryIndexKey,
+				AddressSetIPFamilyID,
+			},
+			EgressServiceOwnerType: {
+				// cluster-wide address set name
+				SecondaryIndexKey,
+				AddressSetIPFamilyID,
+			},
+		},
+	},
+}
+
+// IndexType points to a specific type from dbIndexes map
+type IndexType struct {
+	dbTable         dbObjType
+	contType        controllerType
+	ownerObjectType ownerType
```

Comment:
So we have an index type that is defined by 3 sub-types: dbObjType, controllerType and ownerType. We also have different types for which the set of ids are the same.

I don't have a problem with this but I was asking myself if we could collapse and reduce the number of sub-types and variations for the sake of simplicity.

So right now we have
```
{
	addressSet: {
		defaultController: {
			EgressFirewallDNSOwnerType: {
				// dnsName
				SecondaryIndexKey,
				AddressSetIPFamilyID,
			},
			HybridNodeRouteOwnerType: {
				// nodeName
				SecondaryIndexKey,
				AddressSetIPFamilyID,
			},
			EgressQoSOwnerType: {
				// namespace
				SecondaryIndexKey,
				// priority
				EgressQosAddrSetPriorityKey,
				AddressSetIPFamilyID,
			},
			NetworkPolicyOwnerType: {
				// namespace_name
				SecondaryIndexKey,
				// egress or ingress
				PolicyDirectionID,
				// gress rule index
				GressIdxID,
				AddressSetIPFamilyID,
			},
			NamespaceOwnerType: {
				// namespace
				SecondaryIndexKey,
				AddressSetIPFamilyID,
			},
			EgressIPOwnerType: {
				// cluster-wide address set name
				SecondaryIndexKey,
				AddressSetIPFamilyID,
			},
			EgressServiceOwnerType: {
				// cluster-wide address set name
				SecondaryIndexKey,
				AddressSetIPFamilyID,
			},
		},
	},
}
```
which if we collapsed to remove the concept of controller type and object type, and the ones that are the same, would be equivalent to just:
```
{
	BasicAddressSetType: {
		SecondaryIndexKey,
		AddressSetIPFamilyID,
	},
	EgressQoSAddressSetType: {
		SecondaryIndexKey,
		EgressQosAddrSetPriorityKey,
		AddressSetIPFamilyID,
	},
	NetworkPolicyAddressSetType: {
		SecondaryIndexKey,
		PolicyDirectionID,
		GressIdxID,
		AddressSetIPFamilyID,
	},
}
```
meaning, right now we would not need dbobject type nor controller type as it is not differentiating anything; and for many things we can rely on a BasicAddressSetType.

There is time later on to see how we introduce more variations to this.

Reply:
dbIndex map also provides information about ownerType, which goes directly to the externalIDs. We can specify ownerType in IndexType, or switch back to passing ownertype on dbIndex intialization.
One of the reasons I made this 3-level map is to verify that ids are unique for every object type, .e.g. the same ownerType is not used twice for the dbObjType+controllerType, but it can be reused for other dbObjType or controllerType. So when you introduce a new ownerType, you check which owners already exist under dbIndexes[dbObjType][controllerType], and map won't let add duplicate key.